asteroid/mad.py

```diff
@@ -383,27 +383,80 @@ def _handle_quit(self,_):
-         if args[0].isnumeric(): 
-            self.line_breakpoints.append((int(args[0]),file))
+         self._load_program_text(file)
+         if args[0].isnumeric():
+            if self._validate_breakpoint_line(file, int(args[0])):
+               self.line_breakpoints.append((int(args[0]),file))
          else:
-            self.function_breakpoints.append((args[0],file))
+            if self._validate_breakpoint_function(file, args[0]):
+               self.function_breakpoints.append((args[0],file))
          return START_DEBUGGER
       elif len(args) == 2:
-         if args[0].isnumeric(): 
-            self.line_breakpoints.append((int(args[0]),args[1]))
+         self._load_program_text(args[1])
+         if args[0].isnumeric():
+            if self._validate_breakpoint_line(args[1], int(args[0])):
+               self.line_breakpoints.append((int(args[0]),args[1]))
          else:
-            self.function_breakpoints.append((args[0],args[1]))
+            if self._validate_breakpoint_function(args[1], args[0]):
+               self.function_breakpoints.append((args[0],args[1]))
          return START_DEBUGGER
       else:
          print("error: too many arguments to set")
          return START_DEBUGGER
+   
+   def _validate_breakpoint_line(self, fname, lineno):
+      # Load the correct file
+      file_text = self.program_text[fname]
+      # If the line is outside the current file, produce an error message and return
+      if lineno > len(file_text):
+         print("error: cannot place breakpoint on invalid lines")
+         return False
+      # Check if the line is empty
+      line = file_text[lineno-1].strip()
+      if len(line) == 0:
+         print("error: cannot place breakpoints on blank lines")
+         return False
+      # If the line starts with any of these tokens, it is a valid statement and can accept a breakpoint
+      for start in ['load', 'global', 'structure', 'let', 'loop', 'for', 'while', 'repeat', 'match', 'if', 'try', 'throw', 'break', 'return', 'function']:
+         if len(line) >= len(start) and line[:len(start)] == start:
+            return True
+      # Get the lines above and below the current if they exist
+      prev = file_text[lineno-2].strip() if lineno > 1 else None
+      next = file_text[lineno].strip() if lineno < len(file_text) else None
+      # If the line starts with any of these tokens, the line is in the middle of a statement or expression and cannot accept a breakpoint
+      for start in ['system', 'as', 'with', 'end', 'do', 'in', 'until', 'elif', 'else', 'catch', '@', ',', '[', ']', '(', ')', '"', "'"]:
+         if len(line) >= len(start) and line[:len(start)] == start:
+            print("error: cannot place breakpoint inside a statement")
+            return False
+      # If the previous or next lines begins with any of these symbols, the current line is in the middle of a statement
+      if prev:
+         for start in ['@', ',', '[', '(', '"', "'"]:
```

Comment:
see my comment above, having these collection arbitrarily inside the code makes the code fragile